component-runtime-manager/src/main/java/org/talend/sdk/component/runtime/manager/service/LocalCacheService.java

```diff
@@ -17,71 +17,236 @@
+                (String oldKey , ElementImpl<T> oldElement) -> {
+                    if (oldElement != null && Objects.equals(oldElement.getValue(), expected)) {
+                        // ok to evit, so do release.
+                        oldElement.release();
+                        return null;
+                    }
+                    return oldElement;
+                });
+
     }
 
+
+
     @Override
-    public <T> T computeIfAbsent(final String key, final long timeoutMs, final Supplier<T> value) {
+    public T computeIfAbsent(String key,
+            final Predicate<Element<T>> toRemove,
+            final Supplier<T> value) {
+
+        final ElementImpl<T> element = this.addToMap(key,
+                () -> new ElementImpl<>(value, toRemove, -1, null));
+
+        return element.value;
+    }
+
+    @Override
+    public T computeIfAbsent(String key, long timeoutMs, Supplier<T> value) {
+
+        final long endOfValidity = timeoutMs > 0 ? System.currentTimeMillis() + timeoutMs : -1;
+        final ScheduledFuture<?> task = endOfValidity > 0 ? evictionTask(key, endOfValidity) : null;
+
+        final ElementImpl<T> element = addToMap(key,
+                () -> new ElementImpl<T>(value, null, endOfValidity, task));
+
+        return element.value;
+    }
+
+    private ElementImpl<T> addToMap(String key, Supplier<ElementImpl<T>> builder) {
         final String internalKey = internalKey(key);
-        final Element element = cache
-                .compute(internalKey,
-                        (k, e) -> e == null || e.isExpired()
-                                ? new Element(value.get(), System.currentTimeMillis() + timeoutMs)
-                                : e);
-        if (element == null) {
-            return null;
-        }
-        return (T) element.value;
+        return cache.compute(
+                internalKey,
+                (String k, ElementImpl<T> old) ->    //
+                        old == null || old.mustBeRemoved() ?
+                                builder.get() :
+                                old
+        );
+    }
+
+    @Override
+    public T computeIfAbsent(String key, Supplier<T> value) {
+        final CacheConfiguration config = this.configuration.get();
+        long timeOut = config.isActive() ? config.getDefaultEvictionInterval() : -1;
+        return computeIfAbsent(key, timeOut, value);
     }
 
     private String internalKey(final String key) {
         return plugin + '@' + key;
     }
 
-    @AllArgsConstructor
-    public static class Element {
+    public void clean() {
+        final List<String> removableElements = this.cache
+                .entrySet()
+                .stream()
+                .filter(e -> e.getValue().mustBeRemoved())
+                .map(Entry::getKey)
+                .collect(toList());// materialize before actually removing it
+        removableElements.forEach(this.cache::remove);
+    }
+
+    public boolean isEmpty() {
```

Comment:
not really needed and will impact proxying for nothing IMHO, this is a test thing, just do reflection in the test?

Reply:
removed